### PR TITLE
*: introduce all sql_modes

### PR DESCRIPTION
**Pull request by shenli**
Add a new variable in session to hold the sql_mode.
We will remove StrictSQLMode latter.

See https://github.com/pingcap/tidb/issues/2738 https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html

### CODE REVIEW

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
@tiancaiamao @zimulala PTAL

File: sessionctx/varsutil/varsutil.go [hunk 1]
```diff
@@ -89,11 +89,19 @@ func SetSessionSystemVar(vars *variable.SessionVars, name string, value types.Da
 		vars.TimeZone = parseTimeZone(sVal)
 	case variable.SQLModeVar:
 		sVal = strings.ToUpper(sVal)
+		// TODO: remove this latter.
```
Comment (zimulala) [inline sessionctx/varsutil/varsutil.go]: s/remove/Remove

**Comment by shenli** [comment]
@zimulala PTAL

**Comment by zimulala** [comment]
LGTM

File: mysql/const.go [hunk 2]
```diff
@@ -304,3 +308,90 @@ var DefaultLengthOfTimeFraction = map[int]int{
...
+	str = strings.ToUpper(str)
+	mode, ok := Str2SQLMode[str]
+	if !ok {
+		return ModeNone
+	}
+	return mode
+}
+
+// Str2SQLMode is the string represent of sql_mode to sql_mode map.
+var Str2SQLMode = map[string]SQLMode{
```
Comment (tiancaiamao) [inline mysql/const.go]: use lowercase? `GetSQLMode` is already exposed.

Comment (shenli) [inline mysql/const.go]: I need this for test.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**